### PR TITLE
feat: [CHAOS-715]: Deletion of Environment should not be allowed if any chaos infrastructure is associated with the same.

### PR DESCRIPTION
**Pull request by vanshBhatia-A4k9**
## Describe your changes

- Added 2 new internal APIs, `registerChaosInfrastructure` & `deleteChaosInfrastructure`
- Establishes and removes the entity relationship between chaos infrastructure and environment entity for the chaos module.

Chaos infrastructure inside environment `demo-env`:

Screenshot 2023-09-25 at 5 19 48 PM

Deletion of environment blocked across modules due to entity relationship:

Screenshot 2023-09-25 at 5 20 05 PM

dB entry example:

```
{
  "_id": {
    "$oid": "651173c6016a2961a6244364"
  },
  "accountIdentifier": "kmpySmUISimoRrJL6NL73w",
  "referredEntity": {
    "type": "ENVIRONMENT",
    "entityRef": {
      "scope": "PROJECT",
      "identifier": "demoenv",
      "accountIdentifier": "kmpySmUISimoRrJL6NL73w",
      "orgIdentifier": "default",
      "projectIdentifier": "test",
      "metadata": {},
      "isDefault": true,
      "_class": "io.harness.beans.IdentifierRef"
    },
    "name": ""
  },
  "referredByEntity": {
    "type": "CHAOS_INFRASTRUCTURE",
    "entityRef": {
      "scope": "PROJECT",
      "identifier": "f91e1667-0dc5-4596-a929-360abc1b7f1f",
      "accountIdentifier": "kmpySmUISimoRrJL6NL73w",
      "orgIdentifier": "default",
      "projectIdentifier": "test",
      "metadata": {},
      "isDefault": true,
      "_class": "io.harness.beans.IdentifierRef"
    },
    "name": "demo-infra"
  },
  "referredEntityFQN": "kmpySmUISimoRrJL6NL73w/default/test/demoenv",
  "referredEntityType": "Environment",
  "referredEntityIsDefault": true,
  "referredByEntityFQN": "kmpySmUISimoRrJL6NL73w/default/test/f91e1667-0dc5-4596-a929-360abc1b7f1f",
  "referredByEntityType": "ChaosInfrastructure",
  "referredByEntityIsDefault": true,
  "createdAt": {
    "$numberLong": "1695642566900"
  },
  "lastModifiedAt": {
    "$numberLong": "1695642566900"
  },
  "version": {
    "$numberLong": "0"
  },
  "_class": "io.harness.ng.core.entityReference.entity.EntitySetupUsage"
}
```

Deleting the chaos infrastructure associated with the environment `demo-env`:

Screenshot 2023-09-25 at 5 24 21 PM

Environment can now be safely deleted:

Screenshot 2023-09-25 at 5 24 30 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53289)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/chaos/ChaosResource.java [hunk 1]
```diff
@@ -68,4 +64,26 @@ public ResponseDTO<Boolean> chaosStepNotify(@Body ChaosStepNotifyResponse stepNo
       return ResponseDTO.newResponse(true);
     }
   }
-}
+
+  @POST
+  @Path("/chaosInfrastructure")
+  @InternalApi
+  @ApiOperation(value = "Register the chaos infrastructure entity with the parent environment",
+      nickname = "chaosInfrastructureRegister", hidden = true)
```
Comment (SahilKr24) [inline 120-ng-manager/src/main/java/io/harness/ng/chaos/ChaosResource.java:72]: we can keep this in sync with function name `registerChaosInfrastructure`

File: 120-ng-manager/src/main/java/io/harness/ng/chaos/ChaosResource.java [hunk 2]
```diff
@@ -68,4 +64,26 @@ public ResponseDTO<Boolean> chaosStepNotify(@Body ChaosStepNotifyResponse stepNo
...
+  registerChaosInfrastructure(@Body ChaosInfrastructureRequest chaosInfrastructureRequest) {
+    boolean result = chaosService.registerChaosInfrastructure(chaosInfrastructureRequest);
+    return ResponseDTO.newResponse(result);
+  }
+
+  @DELETE
+  @Path("/chaosInfrastructure")
+  @InternalApi
+  @ApiOperation(value = "Deregister the chaos infrastructure entity relation with the parent environment",
+      nickname = "chaosInfrastructureDelete", hidden = true)
```
Comment (SahilKr24) [inline 120-ng-manager/src/main/java/io/harness/ng/chaos/ChaosResource.java:83]: same here

**Review by SahilKr24 — APPROVED** [review]

**Comment by vanshBhatia-A4k9** [comment]
trigger messagecheck

**Comment by vanshBhatia-A4k9** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/chaos/ChaosServiceImpl.java [hunk 3]
```diff
@@ -6,20 +6,134 @@
...
+      eventProducer.send(
+          Message.newBuilder()
+              .putAllMetadata(ImmutableMap.of("accountId", accountIdentifier,
+                  EventsFrameworkMetadataConstants.REFERRED_ENTITY_TYPE, EntityTypeProtoEnum.ENVIRONMENT.name(),
+                  EventsFrameworkMetadataConstants.ACTION, EventsFrameworkMetadataConstants.FLUSH_CREATE_ACTION))
+              .setData(entityReferenceDTO.toByteString())
+              .build());
+    } catch (Exception ex) {
+      log.info(ENTITY_REFERENCE_LOG_PREFIX
+              + "The entity reference was not created when the chaosInfrastructure [{}] was created in the environment [{}]",
```
Comment (sagarkrsd) [inline 120-ng-manager/src/main/java/io/harness/ng/chaos/ChaosServiceImpl.java:94]: why the message is in past tense here? We could still say
```
The entity reference could not be created...
```

**Review by sagarkrsd — DISMISSED** [review]
LGTM

**Comment by vanshBhatia-A4k9** [comment]
trigger uts

**Comment by vanshBhatia-A4k9** [comment]
trigger ut3

**Comment by vanshBhatia-A4k9** [comment]
trigger uts

**Review by SahilKr24 — APPROVED** [review]

**Pull request merged**